src/main/java/fi/aalto/cs/apluscourses/model/MainViewModelUpdater.java

```diff
@@ -0,0 +1,172 @@
+   * @param project The project to which thsi updater is tied.
+   */
+  public MainViewModelUpdater(@NotNull Project project) {
+    this(PluginSettings.getInstance(), project, PluginSettings.MAIN_VIEW_MODEL_UPDATE_INTERVAL);
+  }
+
+  private void interruptIfProjectClosed() throws InterruptedException {
+    if (aplusProject.getProject().isDisposed() || !aplusProject.getProject().isOpen()) {
+      throw new InterruptedException();
+    }
+  }
+
+  @NotNull
+  URL getCourseUrl() throws InterruptedException {
+    return ReadAction.compute(() -> {
+      interruptIfProjectClosed();
+      try {
+        URL url = aplusProject.getCourseFileUrl();
+        if (url == null) {
+          throw new InterruptedException();
+        }
+        return url;
+      } catch (IOException e) {
+        throw new InterruptedException();
+      }
+    });
+  }
+
+  @Nullable
+  Course getCourse(@NotNull URL courseUrl) {
+    // We don't need to be inside the read action when we parse the course configuration file.
+    // This way we avoid long read actions, which block the UI thread from making write
+    // actions and lead to UI freezes. We just have to remember to later check again that the
+    // project is still actually open.
+    try {
+      return Course.fromUrl(courseUrl, new IntelliJModelFactory(aplusProject.getProject()));
+    } catch (IOException | MalformedCourseConfigurationFileException e) {
+      return null;
+    }
+  }
+
+  @NotNull
+  Set<String> getProjectModules() throws InterruptedException {
+    com.intellij.openapi.module.Module[] projectModules = ReadAction.compute(() -> {
+      interruptIfProjectClosed();
+      return aplusProject.getModuleManager().getModules();
+    });
+    return Arrays.stream(projectModules)
+        .map(module -> module.getName())
+        .collect(Collectors.toSet());
+  }
+
+  @NotNull
+  List<Module> getUpdatableModules(@NotNull Course course) throws InterruptedException {
+    List<Module> updatableModules = new ArrayList<>();
+
+    // Modules listed in the course file may have been removed from the project, so we have to check
+    // that the modules are still in the project.
+    Set<String> projectModules = getProjectModules();
+
+    // Minor concurrency issue: we make the list of updatable modules outside of a read action, so
+    // (however unlikely) a module may be removed from the project in the meantime. This shouldn't
+    // be a big issue however, as it would just lead to a update notification for a module that has
+    // been removed.
+
+    Map<String, String> localModuleIds;
+    try {
+      localModuleIds = aplusProject.getCourseFileModuleIds();
+    } catch (IOException e) {
+      return updatableModules;
+    }
+
+    for (Module module : course.getModules()) {
+      // An updatable module must be in the project and it's ID in the local
+      // course file must be different from the ID in the course configuration file.
```

Comment:
but why? @nikke234

Reply:
If the version IDs are equal in local and remote, then there's no updates, so the module is not updatable. Would it be clearer if the wording was "To be updatable, a module must be in the project an its ID [...] must be different from [...]"?

Reply:
Yes that's a clearer wording :+1:

Reply:
I've totally forgot about the IDs changing as well, sorry :bow: